kornia/filters/guided.py

```diff
@@ -0,0 +1,217 @@
+        kernel_size: the size of the kernel.
+        eps: regularization parameter. Smaller values preserve more edges.
+        border_type: the padding mode to be applied before convolving.
+          The expected modes are: ``'constant'``, ``'reflect'``,
+          ``'replicate'`` or ``'circular'``. Default: ``'reflect'``.
+        subsample: subsampling factor for Fast Guided filtering. Default: 1 (no subsampling)
+
+    Returns:
+        the blurred tensor with same shape as `input` :math:`(B, C, H, W)`.
+
+    Examples:
+        >>> guidance = torch.rand(2, 3, 5, 5)
+        >>> input = torch.rand(2, 4, 5, 5)
+        >>> output = guided_blur(guidance, input, 3, 0.1)
+        >>> output.shape
+        torch.Size([2, 4, 5, 5])
+    """
+
+    KORNIA_CHECK_IS_TENSOR(guidance)
+    KORNIA_CHECK_SHAPE(guidance, ['B', 'C', 'H', 'W'])
+    if input is not guidance:
+        KORNIA_CHECK_IS_TENSOR(input)
+        KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+        KORNIA_CHECK(
+            (guidance.shape[0] == input.shape[0]) and (guidance.shape[-2:] == input.shape[-2:]),
+            "guidance and input should have the same batch size and spatial dimensions",
+        )
+
+    kernel_size = _unpack_2d_ks(kernel_size)
+    if guidance.shape[1] == 1:
+        return _guided_blur_grayscale_guidance(guidance, input, kernel_size, eps, border_type, subsample)
+    else:
+        return _guided_blur_multichannel_guidance(guidance, input, kernel_size, eps, border_type, subsample)
+
+
+class GuidedBlur(Module):
+    r"""Blur a tensor using a Guided filter.
+
+    The operator is an edge-preserving image smoothing filter. See :cite:`he2010guided`
+    and :cite:`he2015fast` for details. Guidance and input can have different number of channels.
+
+    Arguments:
+        kernel_size: the size of the kernel.
+        eps: regularization parameter. Smaller values preserve more edges.
+        border_type: the padding mode to be applied before convolving.
+          The expected modes are: ``'constant'``, ``'reflect'``,
+          ``'replicate'`` or ``'circular'``. Default: ``'reflect'``.
+        subsample: subsampling factor for Fast Guided filtering. Default: 1 (no subsampling)
+
+    Returns:
+        the blurred input tensor.
+
+    Shape:
+        - Input: :math:`(B, C, H, W)`, :math:`(B, C, H, W)`
+        - Output: :math:`(B, C, H, W)`
+
+    Examples:
+        >>> guidance = torch.rand(2, 3, 5, 5)
+        >>> input = torch.rand(2, 4, 5, 5)
+        >>> blur = GuidedBlur(3, 0.1)
+        >>> output = blur(guidance, input)
+        >>> output.shape
+        torch.Size([2, 4, 5, 5])
+    """
+
+    def __init__(
+        self, kernel_size: tuple[int, int] | int, eps: float | Tensor, border_type: str = 'reflect', subsample: int = 1
+    ):
+        super().__init__()
+        self.kernel_size = kernel_size
+        if isinstance(eps, Tensor):
+            self.register_buffer("eps", eps)
+        else:
+            self.eps = eps
```

Comment:
```suggestion
        self.register_buffer("eps", as_tensor(eps))
```

Reply:
There is another case, when `eps` is trainable, we need to use `self.register_parameter()`. How should we handle it? I saw one approach here #1923, but it doesn't feel convenient to use, as the user needs to wrap `eps` in `nn.Parameter()`. How about having an argument `trainable_eps` in `__init__()`?

Reply:
has any gain in the eps been a Tensor instead of a simple float here?

Reply:
From what I see, GaussianBlur also accepts sigma as Tensor so I follow it

https://github.com/kornia/kornia/blob/0ff08c72d5b5da101f1e57916cba892c7b3a2905/kornia/filters/gaussian.py#L104-L110

You are right, for non-trainable `eps`, there is no benefit to have it as a Tensor. I will make `eps` as `float` only. (Perhaps we should update other modules to not accept tensor also? like sigma in GaussianBlur)

What do you think about having `trainable_eps` as an argument in `__init__()`? To have `eps` as trainable parameter. (Then we will do something like `self.register_parameter("eps", nn.Parameter(torch.tensor([eps])))`)